src/SmallStrainPlasticity/SmallStrainPlasticity.jl

```diff
@@ -0,0 +1,200 @@
+= \\dot{\\lambda} \\boldsymbol{\\nu}
+```
+
+The plastic multiplier, ``\\lambda``, is obtained via the KKT-conditions,
+```math
+\\dot{\\lambda} \\geq 0, \\quad \\Phi \\leq 0, \\quad \\dot{\\lambda}\\Phi = 0
+```
+
+The isotropic hardening is formulated as
+```math
+\\kappa = \\sum_{i=1}^{N_{\\mathrm{iso}}} g_{\\mathrm{iso},i}(\\lambda)
+```
+where ``g_{\\mathrm{iso},i}(\\lambda)`` is specified by `m.isotropic[i]`
+
+Kinematic hardening is formulated as
+```math
+\\boldsymbol{\\beta}_i = \\dot{\\lambda} g_{\\mathrm{kin},i}(\\nu, \\boldsymbol{\\beta}_i)
+```
+where ``g_{\\mathrm{kin},i}(\\boldsymbol{\\nu}, \\boldsymbol{\\beta}_i)`` is specified by `m.kinematic[i]`
+and ``i\\in[1,N_\\mathrm{kin}]``. 
+
+# Algorithmic tangent stiffness computations
+σ(X(ϵ), ϵ): dσdϵ = ∂σ∂X dXdϵ + ∂σ∂ϵ
+R(X(ϵ), ϵ): dRdX=0=∂R∂X dXdϵ + ∂R∂ϵ
+dσdϵ = - ∂σ∂X [∂R∂X]^-1 ∂R∂ϵ + ∂σ∂ϵ
+
+For this specific case, 
+
+- ∂σ∂X is identity in upper left corner, zero the rest
+- ∂R∂ϵ is the elastic stiffness in upper left corner, zero the rest
+- ∂σ∂ϵ is zero (the entire stress σ is only a function of x.σ)
+
+# Keyword arguments
+- `cache`: Cache for the iterative solver, used by NLsolve.jl. It is strongly recommended to pre-allocate the cache for repeated calls to `material_response`. See [`get_cache`](@ref).
+- `options::Dict{Symbol, Any}`: Solver options for the non-linear solver. Under the key `:nlsolve_params` keyword arguments for `nlsolve` can be handed over.
+See [NLsolve documentation](https://github.com/JuliaNLSolvers/NLsolve.jl#common-options). By default the Newton solver will be used.
+"""
+function material_response(m::Chaboche, ϵ::SymmetricTensor{2,3}, old::ChabocheState{Nkin,T,N}, Δt; cache=get_cache(m), options::Dict{Symbol, Any} = Dict{Symbol, Any}()) where {T,N,Nkin}
+    
+    σ_trial, dσdϵ_elastic, _, _ = material_response(m.elastic, ϵ-old.ϵₚ, nothing, Δt)
+    Φ_trial = vonmises(σ_trial-sum(old.β)) - (m.σ_y0 + sum(get_hardening.(m.isotropic, old.λ)))
+
+    if Φ_trial < 0
+        return σ_trial, dσdϵ_elastic, old, true
+    else
+        x0 = initial_guess(m, old, ϵ)
+        rf!(r_vector, x_vector) = vector_residual!((x)->residual(x,m,old,ϵ), r_vector, x_vector, x0)  # Using x0 as template for residual instead of material as this is related to Tensors
+        update_cache!(cache.R_X_oncediff, rf!)
+        
+        tomandel!(cache.R_X_oncediff.x_f, x0)
+        # Should this be centrally managed? I.e. process_options or similar?
+        nlsolve_options = get(options, :nlsolve_params, Dict{Symbol, Any}(:method=>:newton))
+        haskey(nlsolve_options, :method) || merge!(nlsolve_options, Dict{Symbol, Any}(:method=>:newton)) # set newton if the user did not supply another method
+        
+        # Solve local problem:
+        result = NLsolve.nlsolve(cache.R_X_oncediff, cache.R_X_oncediff.x_f; nlsolve_options...)
+
+        if result.f_converged
+            x = frommandel(ChabocheResidual{Nkin}, result.zero)
```

Comment:
```suggestion
            x = frommandel(ChabocheResidual{Nkin}, result.zero::Vector{T})
```
I think you'll have to annotate this for type stability. It's the nasty problem when handing a kwarg-dict to `nlsolve`.

Reply:
Nice, this solves the type instabilities!

As we discussed, extracting directly from the OnceDifferentiable by field name in the material is not so nice. However, would it be a nicer solution if we have functions such as these in a general utility routine 
```
get_oncediff_result(oncediff::OnceDifferentiable) = oncediff.x_f
get_oncediff_jacobian(oncediff::OnceDifferentiable) = oncediff.DF
```
and then use e.g. `x = frommandel(ChabocheResidual{Nkin}, get_oncediff_result(cache.R_X_oncediff)` ?
To me this seems neater than forcing the type in each material (this also makes it easier to switch between e.g. Vector and MVector....)